src/MongoDB.Driver/MongoClient.cs

```diff
@@ -338,7 +338,11 @@ private bool AreSessionsSupported(CancellationToken cancellationToken)
 
         private bool? AreSessionsSupported(ClusterDescription clusterDescription)
         {
-            if (clusterDescription.Servers.Any(s => s.IsDataBearing))
+            var isDirectConnection = clusterDescription.ConnectionMode == ClusterConnectionMode.Direct;
+            var connectedServers = clusterDescription.Servers.Where(s => s.State == ServerState.Connected);
+            var pertinentServers = isDirectConnection ? connectedServers : connectedServers.Where(s => s.IsDataBearing);
+
+            if (pertinentServers.Any())
```

Comment:
This and the change below are the only two changes to the driver.

The rest was work in the test projects to get to a place where these changes could be tested.

Reply:
This duplicates the logic in AreSessionsSupportedServerSelector.  Can we DRY it up?

Reply:
Yes.

Centralized logic for determining whether sessions are supported in ClusterDescription.